cmd/swarm/main.go

```diff
@@ -333,6 +348,43 @@ func bzzd(ctx *cli.Context) error {
 	return nil
 }
 
+// newEnsClient creates a new Ethereum API client which is used to resolve ENS
+// names, and also determines the ENS contract address by getting both the
+// version and genesis hash using the client and matching them to either
+// mainnet or testnet addresses
+func newEnsClient(url string) (*ethclient.Client, common.Address, error) {
+	client, err := rpc.Dial(url)
+	if err != nil {
+		return nil, common.Address{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	var version string
+	if err := client.CallContext(ctx, &version, "net_version"); err != nil {
+		client.Close()
+		return nil, common.Address{}, err
+	}
+
+	ethClient := ethclient.NewClient(client)
+	block, err := ethClient.BlockByNumber(ctx, big.NewInt(0))
+	if err != nil {
+		client.Close()
+		return nil, common.Address{}, err
+	}
+
+	var contractAddr common.Address
```

Comment:
Are you talking about a custom address coming from the on-disk config? If so, we will need to do this detection earlier and pass it to `NewConfig`, otherwise we don't know if `NewConfig` used a default address (i.e. the testnet address), or an explicit one.